gdnative-core/src/globalscope.rs

```diff
@@ -0,0 +1,395 @@
+pub fn smoothstep(from: f32, to: f32, s: f32) -> f32 {
+    if is_equal_approx(from, to) {
+        return from;
+    }
+    let s = ((s - from) / (to - from)).clamp(0.0, 1.0);
+    s * s * (3.0 - 2.0 * s)
+}
+/// Returns `true` if `a` and `b` are approximately equal to each other.
+/// Here, approximately equal means that `a` and `sb` are within a small internal epsilon of each other,
+/// which scales with the magnitude of the numbers.
+/// Infinity values of the same sign are considered equal.
+#[inline]
+pub fn is_equal_approx(a: f32, b: f32) -> bool {
+    if a == b {
+        return true;
+    }
+    let mut tolerance = CMP_EPSILON * a.abs();
+    if tolerance < CMP_EPSILON {
+        tolerance = CMP_EPSILON;
+    }
+    (a - b).abs() < tolerance
+}
+
+/// Returns true if s is zero or almost zero.
+/// This method is faster than using is_equal_approx with one value as zero.
+#[inline]
+pub fn is_zero_approx(s: f32) -> bool {
+    s.abs() < CMP_EPSILON
+}
+/// Converts from linear energy to decibels (audio).
+/// This can be used to implement volume sliders that behave as expected (since volume isn't linear).
+#[inline]
+pub fn linear2db(nrg: f32) -> f32 {
+    nrg.ln() * 0.115_129_255
+}
+/// Returns the nearest equal or larger power of 2 for integer value.
+/// In other words, returns the smallest value a where `a = pow(2, n)` such that `value <= a` for some non-negative integer `n`.
+/// # Examples:
+/// ```rust
+/// use gdnative_core::globalscope::nearest_po2;
+/// assert_eq!(nearest_po2(3), 4);
+/// assert_eq!(nearest_po2(4), 4);
+/// assert_eq!(nearest_po2(5), 8);
+/// assert_eq!(nearest_po2(0), 0);
+/// assert_eq!(nearest_po2(-1), 0);
+/// ```
+#[inline]
+pub fn nearest_po2(value: i32) -> u32 {
+    if value <= 0 {
+        return 0;
+    }
+    (value as u32).next_power_of_two()
+}
+/// Converts a 2D point expressed in the polar coordinate system
+/// (a distance from the origin r and an angle th (radians)) to the cartesian coordinate system (X and Y axis).
+#[inline]
+pub fn polar2cartesian(r: f32, th: f32) -> Vector2 {
+    Vector2::new(r * th.cos(), r * th.sin())
+}
+/// Returns the integer modulus of a/b that wraps equally in positive and negative.
+/// # Examples:
+/// ```rust
+/// use gdnative_core::globalscope::posmod;
+/// const VALS: [i32; 7] = [0, 1, 2, 0, 1, 2, 0];
+/// for i in (-3..4).enumerate() {
+///     assert_eq!(posmod(i.1, 3), VALS[i.0]);
+/// }
+/// ```
+#[inline]
+pub fn posmod(a: i32, b: i32) -> i32 {
+    let mut value = a % b;
+    if ((value < 0) && (b > 0)) || ((value > 0) && (b < 0)) {
+        value += b;
+    }
+    value
+}
+/// Maps a value from range [istart, istop] to [ostart, ostop].
+/// # Example:
+/// ```rust
+/// use gdnative_core::globalscope::range_lerp;
+/// assert_eq!(range_lerp(75.0, 0.0, 100.0, -1.0, 1.0), 0.5);
+/// ```
+#[inline]
+pub fn range_lerp(value: f32, istart: f32, istop: f32, ostart: f32, ostop: f32) -> f32 {
+    lerp(ostart, ostop, inverse_lerp(istart, istop, value))
+}
```

Comment:
With the suggested `Range` (see main comment), this could be:
```suggestion
/// Maps a value from range [istart, istop] to [ostart, ostop].
/// # Example:
/// ```rust
/// use gdnative::globalscope::range_lerp;
/// assert_eq!(range_lerp(75.0, 0.0..100.0, -1.0..1.0), 0.5);
/// ```
#[inline]
pub fn range_lerp(value: f32, from: Range<f32>, to: Range<f32>) -> f32 {
    lerp(to, inverse_lerp(from, value))
}
```

Reply:
thanks for the review i will work on these things :)